thavalon-server/src/database/accounts/mod.rs

```diff
@@ -0,0 +1,172 @@
+pub mod account_errors;
+use super::get_db_client;
+use account_errors::AccountError;
+use mongodb::{
+    bson::{self, doc},
+    options::FindOneOptions,
+};
+use serde::{Deserialize, Serialize};
+const USER_COLLECTION: &str = "thavalon_users";
+/// Canonical representation of a database account.
+/// This should never leave the database, as it contains a password hash!
+#[derive(Serialize, Deserialize)]
+pub struct DatabaseAccount {
+    pub email: String,
+    pub hash: String,
+    pub display_name: String,
+    pub profile_picture: Option<Vec<u8>>,
+    pub email_verified: bool,
+}
+
+/// Creates a new user and adds to the database
+///
+/// # Arguments
+///
+/// * `user` - A DatabaseAccount holding information for the new user
+///
+/// # Returns
+///
+/// * Null on success, AccountCreationError on failure
+pub async fn create_new_user(user: &DatabaseAccount) -> Result<(), AccountError> {
+    log::info!("Attempting to add thavalon user: {}.", user.email);
+    if does_user_exist(user).await {
```

Comment:
Can we push this check into the mongo query? Otherwise it’s race-condition-prone

Reply:
I got rid of that function entirely. Now we use one Mongo call to insert the document and error if we find a matching document (i.e. the user account already exists). I *think* I set up the query so we don't override the first user's account  in this case :).